R/db2.R

```diff
@@ -459,9 +470,46 @@ report_db2_destroy <- function(con) {
 }
 
 
-report_db2_publish <- function(con, id, value) {
+report_db2_publish <- function(con, id, name, value) {
   sql <- "UPDATE report_version SET published = $1 WHERE id = $2"
   DBI::dbExecute(con, sql, list(value, id))
+  report_db2_update_changelog_published(con, name)
+}
+
+
+report_db2_update_changelog_published <- function(con, name) {
+  sql <- "SELECT id FROM report_version WHERE report = $1 and published"
+  published <- DBI::dbGetQuery(con, sql, name)$id
+
+  sql <- paste("SELECT",
+               "  changelog.id, report_version, report_version_public,",
+               "    published",
+               "  FROM changelog",
+               "  JOIN report_version",
+               "    ON report_version.id = changelog.report_version",
+               "  JOIN changelog_label",
+               "    ON changelog_label.id = changelog.label",
+               " WHERE report_version.report = $1",
+               "   AND changelog_label.public",
+               " ORDER BY report_version")
+  dat <- DBI::dbGetQuery(con, sql, list(name))
+  dat$published <- dat$published == 1
+
+  p <- rep(NA_character_, nrow(dat))
+  for (i in seq_len(nrow(dat))) {
+    j <- dat$report_version[[i]] <= published
+    if (!any(j)) {
+      break
+    }
+    p[[i]] <- published[[min(which(j))]]
+  }
+
+  prev <- dat$report_version_public
+  update <- is.na(p) != is.na(prev) | (!is.na(p) & !is.na(prev) & p != prev)
```

Comment:
I don't follow the the logic here. Unless R allowed to short circuit expressions like this, I think this should evaluate to `update <- is.na(p)`.

And I'm not 100% sure that `&` and `|` are higher precedence than `!=`.

There is probably a clearer way of expressing the logic in this line.

Reply:
There is no short circuiting in here, but the logic does not expand as you suggest.  `!=` binds more tightly than `&`

The logic here is the following

|old|new|updated?|
|---|---|---|
|NA|NA|FALSE|
|x|x|FALSE|
|NA|x|TRUE|
|x|NA|TRUE|
|x|y|TRUE|

so we might be able to do it with:

```r
!(is.na(p) & is.na(prev) | (!is.na(p) & !is.na(prev) & p == prev))
```

but that's about as much logic.  What do you think?